File: examples/backend/php/UserServiceTest.php
<?php

declare(strict_types=1);

namespace App\User;

use InvalidArgumentException;
use PHPUnit\Framework\TestCase;

/**
 * Unit test with an in-memory fake — no framework bootstrap, no database.
 */
final class InMemoryUserRepository implements UserRepository
{
    /** @var array<int, UserDTO> */
    private array $users = [];
    private int $nextId = 1;

    public function findById(int $id): ?UserDTO
    {
        return $this->users[$id] ?? null;
    }

    public function findByUsername(string $username): ?UserDTO
    {
        foreach ($this->users as $user) {
            if ($user->username === $username) {
                return $user;
            }
        }

        return null;
    }

    public function insert(string $username, string $email): UserDTO
    {
        $user = new UserDTO($this->nextId, $username, $email);
        $this->users[$user->id] = $user;
        $this->nextId++;

        return $user;
    }
}

final class UserServiceTest extends TestCase
{
    private UserService $service;

    #[\Override]
    protected function setUp(): void
    {
        $this->service = new UserService(new InMemoryUserRepository());
    }

    public function testRegisterCreatesUser(): void
    {
        $user = $this->service->register('ada', 'ada@example.com');

        $this->assertSame('ada', $user->username);
        $this->assertSame(1, $user->id);
    }

    public function testRegisterRejectsDuplicateUsername(): void
    {
        $this->service->register('ada', 'ada@example.com');

        $this->expectException(InvalidArgumentException::class);
        $this->expectExceptionMessage('already taken');
        $this->service->register('ada', 'other@example.com');
    }

    public function testRegisterRejectsInvalidEmail(): void
    {
        $this->expectException(InvalidArgumentException::class);
        $this->service->register('ada', 'not-an-email');
    }

    public function testGetUserThrowsWhenMissing(): void
    {
        $this->expectException(UserNotFoundException::class);
        $this->service->getUser(999);
    }
}

File: examples/backend/java/UserServiceTest.java
package com.example.user;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.when;

import com.example.user.UserService.UserDTO;
import com.example.user.UserService.UserNotFoundException;
import com.example.user.UserService.UserRepository;
import java.util.Optional;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

/**
 * Unit test — no Spring context loaded. Behavior-describing names.
 */
@ExtendWith(MockitoExtension.class)
class UserServiceTest {

    @Mock
    private UserRepository repository;

    private UserService service;

    @BeforeEach
    void setUp() {
        service = new UserService(repository);
    }

    @Test
    void shouldReturnUserWhenUserExists() {
        when(repository.findById(1L)).thenReturn(Optional.of(new UserDTO(1L, "ada", "ada@example.com")));

        UserDTO user = service.getUser(1L);

        assertThat(user.username()).isEqualTo("ada");
    }

    @Test
    void shouldThrowNotFoundWhenUserDoesNotExist() {
        when(repository.findById(999L)).thenReturn(Optional.empty());

        assertThatThrownBy(() -> service.getUser(999L))
                .isInstanceOf(UserNotFoundException.class)
                .hasMessageContaining("999");
    }

    @Test
    void shouldRejectDuplicateUsername() {
        when(repository.findByUsername("ada")).thenReturn(Optional.of(new UserDTO(1L, "ada", "ada@example.com")));

        assertThatThrownBy(() -> service.register("ada", "other@example.com"))
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessageContaining("already taken");
    }

    @Test
    void shouldRejectInvalidEmail() {
        assertThatThrownBy(() -> service.register("ada", "not-an-email"))
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessageContaining("invalid email");
    }

    @Test
    void shouldInsertWhenUsernameIsAvailable() {
        when(repository.findByUsername(anyString())).thenReturn(Optional.empty());
        when(repository.insert("ada", "ada@example.com")).thenReturn(new UserDTO(1L, "ada", "ada@example.com"));

        UserDTO created = service.register("ada", "ada@example.com");

        assertThat(created.id()).isEqualTo(1L);
    }
}
